README.md
# SVG Logo Generator

[![License Badge](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

## Description
This node.js application will build SVG Logos on demand. Logos can be in the shape of a circle, square or a triangle and contain between 1 and 3 characters of text. Logos and text can be of any CSS colour.

## Table of Contents
* [Video Demonstration](#video-demonstration)
* [Installation](#installation)
* [Usage](#usage)
* [License](#license)
* [Contributing](#contributing)
* [Tests](#tests)
* [Questions](#questions)

## Video Demonstration
Click the link to view the [Demonstration Video][def1]


![Demonstration Video](assets/video/svgLogoGenerator.gif)
## Installation
The SVG Logo Generator requires node.js & Inquirer v8.2.4 to be installed. The application also uses Jest v24.9.0 to confirm functionality. You can clone the package from the GitHub Repository. Install Inquirer & Jest by running npm install from the CLI.


## Usage
The Project is started from the CLI by running 'node index.js'. You will be prompted to answer a series of questions. The answers will be used to build the SVG Logo file (logo.svg) and save it in the examples folder.

## License
REAME Generator is available under the MIT License license. See the [LICENSE](https://opensource.org/licenses/MIT) file for more info.

## Contributing
Contributors are welcome. Feel free to clone or fork the project. Please let me know how you get on.

## Tests
Automated testing is provided to ensure the Logo Generator is working correctly. Jest v24.9.0 is used for testing. The test script is shapes.test.js located in the lib folder. shapes.test.js attempts to generate each of the 3 provided shapes (circle, square & triangle) and report success or any arrors. You can confirm functionality from the CLI using this command: 'npm run test'. 


## Questions
  
If you have questions about this project or would like more information you can [contact me by email](mailto:peter.medbury@dingogap.net.au).
  
You can see more of my work in my [Github Repository](https://github.com/dingogap).

## Resources Used
1. Bootcamp Materials - Lesson Material from Weeks 1 - 10
2. Prior knowledge and experience
3. [Inquirer.js][def2] Documentation
4. [Jest][def3] Documentation
5. [SVG Tutorial][def4] on MDM
6. [Node.js Exports Shortcut][def5] Documentation


[def1]: https://vimeo.com/849850517
[def2]: https://www.npmjs.com/package/inquirer
[def3]: https://jestjs.io/docs/29.4/getting-started
[def4]: https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorial/Getting_Started
[def5]: https://nodejs.org/api/modules.html#exports-shortcut

CHANGELOG.md
# SVG Logo Maker

1. Initialised GitHub Repository
2. Created Folder Structure
3. Initialised npm & installed inquirer & jest packages
4. Created questions.js in the lib folder and created the questions array
5. Created buildLogo.js in the lib folder to hold code to control logo build
6. Add code to ask questions and write the Logo to file
7. Added shape classes and code to render logos
8. Parameterised Triangle polygon points
9. added & tested Jest test script, testing each log shape was generated correctly
10. Testing ways to derive logo positioning
11. Looking at ways to make SVG a class rather than just functions
12. Decided to leave as is for simplicity.
13. Confirmed that Logo size & position could be generated automatically knowing the container size and the logo diameter  


lib/shapes.test.js
// Logo Test Code to be run using Jest
const Circle = require('./circle.js');
const Square = require('./square.js');
const Triangle = require('./triangle.js');

// Checks to see that the output from the render method produces the expectd output - the code to render the circle logo
describe('Circle', () => {
    it('should generate a Circle Logo with colour: blue; text: PDM & textColour: white', () => {
        const circle = new Circle('grey', 'PDM', 'gold');
        expect(circle.render()).toEqual(`<svg version="1.1" width="300" height="200" xmlns="http://www.w3.org/2000/svg">
    <circle cx="150" cy="100" r="80" fill="grey" />
    <text x="150" y="125" font-size="60" text-anchor="middle" fill="gold">PDM</text>
</svg>`
        );
    });
});

// Checks to see that the output from the render method produces the expectd output - the code to render the square logo
describe('Square', () => {
    it('should generate a Square Logo with colour: blue; text: PDM & textColour: white', () => {
        const square = new Square('grey', 'PDM', 'gold');
        expect(square.render()).toEqual(`<svg version="1.1" width="300" height="200" xmlns="http://www.w3.org/2000/svg">
    <rect x="70" y="20" width="160" height="160" fill="grey"/>
    <text x="150" y="125" font-size="60" text-anchor="middle" fill="gold">PDM</text>
</svg>`
        );
    });
});

// Checks to see that the output from the render method produces the expected output - the code to render the triangle logo
describe('Triangle', () => {
    it('should generate a Triangle Logo with colour: blue; text: PDM & textColour: white', () => {
        const triangle = new Triangle('grey', 'PDM', 'gold');
        expect(triangle.render()).toEqual(`<svg version="1.1" width="300" height="200" xmlns="http://www.w3.org/2000/svg">
    <polygon points="95, 20 95, 180 255, 100" fill="grey"/>
    <text x="150" y="125" font-size="60" text-anchor="middle" fill="gold">PDM</text>
</svg>`
        );
    });
});


index.js
// Reference Packages
const inquirer = require("inquirer");
const fs = require("fs");
const logoDefintion = require('./lib/svgStructure.js')

// Include external js files
const questions = require('./lib/questions.js');
const buildLogo = require('./lib/buildLogo.js');

// Declare variables
const fileName = "./examples/logo.svg"

// Start Processing
init();

// Intialise Processing - Ask Questions & when answered generate the logo code & write it to file 
function init() {
    inquirer
        .prompt(questions)
        .then((answers) => {
            return buildLogo(answers);
        })
        .then((data) => {
            writeToFile(fileName, data);
        })
        .catch((error) => {
            console.log(error);
        });
}

// Create file filename & write logo data to it
// Log success to console
function writeToFile(fileName, data) {
    fs.writeFile(fileName, data, (err) => {
        if (err) {
            console.error(err);
        } else {
            console.log(`${fileName} created successfully`);
        }
    });
}

lib/circle.js
// Create a Circle Logo
const Shape = require('./shape.js')
const svgStructure = require('./svgStructure.js')

class Circle extends Shape {
    constructor(colour, text, textColour) {
        super(colour, text, textColour);
        this.colour = colour;
        this.text = text;
        this.textColour = textColour;
    };

    render() {
        // SVG width and height parameters
        const width = "300";
        const height = "200";

        // Circle Parameters: position & radius
        const cx = "150";
        const cy = "100";
        const r = "80";

        // Settings for SVG Text: position, font size & anchor, colour & content
        const x = "150";
        const y = "125";
        const fontSize = "60";
        const textAnchor = "middle";

        // build the SVG logo
        return svgStructure.svgOpenTag(width, height)
            + svgStructure.svgCircle(cx, cy, r, this.colour)
            + svgStructure.svgText(x, y, fontSize, textAnchor, this.textColour, this.text)
            + svgStructure.svgCloseTag()
    };
};

module.exports = Circle;
  

lib/square.js
// Creare a Square Logo
const Shape = require('./shape.js')
const svgStructure = require('./svgStructure.js')

class Square extends Shape {
    constructor(colour, text, textColour) {
        super(colour, text, textColour);
        this.colour = colour;
        this.text = text;
        this.textColour = textColour;
    };

    render() {
        // SVG width and height parameters
        const width = "300";
        const height = "200";

        // Circle Parameters: position & radius
        const cx = "70";
        const cy = "20";
        const rectangleWidth = "160";
        const rectangleHeight = "160";

        // Settings for SVG Text: position, font size & anchor, colour & content
        const x = "150";
        const y = "125";
        const fontSize = "60";
        const textAnchor = "middle";
        return svgStructure.svgOpenTag(width, height)
            + svgStructure.svgSquare(cx, cy,rectangleWidth, rectangleHeight, this.colour)
            + svgStructure.svgText(x, y, fontSize, textAnchor, this.textColour, this.text)
            + svgStructure.svgCloseTag()
    };
};

module.exports = Square;

lib/questions.js
// Prompt for logo text, text colour, logo shape & logo colour
const questions = [
    {
        name: "logoText",
        type: "input",
        message: "Enter Logo Text (at least 1 character & no more than 3 characters):",
        validate: (answer) => {
            if (answer.length > 3 || answer.length === 0) {
                return console.log("Please enter the Text for the Logo (at least 1 character & no more than 3 characters)");
            }
            return true;
        },
    },
    {
        name: "logoTextColour",
        type: "input",
        message: "Enter Logo Text Colour (colour keyword or hexadecimal number)",
        validate: (answer) => {
            if (answer.length === 0) {
                return console.log("Please enter the Colour of the Logo Text:");
            } else {
                return true;
            }
        }
    },
    {
        name: "logoShape",
        type: "list",
        message: "Enter the shape of the Logo",
        choices: ["Circle", "Square", "Triangle"],
    },
    {
        name: "logoColour",
        type: "input",
        message: "Enter Logo Colour (colour keyword or hexadecimal number)",
        validate: (answer) => {
            if (answer.length === 0) {
                return console.log("Please enter the Colour of the Logo:");
            } else {
                return true;
            }
        }
    },
]
module.exports = questions;

lib/svgStructure.js
// Generic code to generate SVG Logo Shapes. This code will not need to be changed if dimensions change

// SVG definition
exports.svgOpenTag = (width, height) => {
    return `<svg version="1.1" width="${width}" height="${height}" xmlns="http://www.w3.org/2000/svg">`
}
// Generate Circle
exports.svgCircle = (cx,cy,r,logoColour) => {
    return `
    <circle cx="${cx}" cy="${cy}" r="${r}" fill="${logoColour}" />`
}

// Generate Square
exports.svgSquare = (x, y, width, height, logoColour) => {
    return `
    <rect x="${x}" y="${y}" width="${width}" height="${height}" fill="${logoColour}"/>`
}

// Generate Triangle
exports.svgTriangle = (x1, y1, x2, y2, x3, y3,logoColour) => {
    return `
    <polygon points="${x1}, ${y1} ${x2}, ${y2} ${x3}, ${y3}" fill="${logoColour}"/>`
}

// SVG Text tag: position, font size, anchor, colour & content
exports.svgText = (x,y,fs,ta,textColour,text) => {
    return `
    <text x="${x}" y="${y}" font-size="${fs}" text-anchor="${ta}" fill="${textColour}">${text}</text>`
}

// SVG closing tag
exports.svgCloseTag = () => {
    return `
</svg>`
}



lib/triangle.js
// Create a Triangle Logo
const Shape = require('./shape.js')
const svgStructure = require('./svgStructure.js')

class Triangle extends Shape {
    constructor(colour, text, textColour) {
        super(colour, text, textColour);
        this.colour = colour;
        this.text = text;
        this.textColour = textColour;
    };

    render() {
        // SVG width and height parameters
        const width = "300";
        const height = "200";

        // Polygon point placement to create Triangle
        const x1 = "95"
        const y1 = "20"
        const x2 = "95"
        const y2 = "180"
        const x3 = "255"
        const y3 = "100"

        // Settings for SVG Text: position, font size & anchor, colour & content
        const x = "150";
        const y = "125";
        const fontSize = "60";
        const textAnchor = "middle";

        return svgStructure.svgOpenTag(width, height)
            + svgStructure.svgTriangle(x1, y1, x2, y2, x3, y3, this.colour)
            + svgStructure.svgText(x, y, fontSize, textAnchor, this.textColour, this.text)
            + svgStructure.svgCloseTag()
    };
};

module.exports = Triangle;
 